internal/scraper/scraper.go

```diff
@@ -427,7 +383,116 @@ func (s Scraper) collectData(ctx context.Context, t time.Time) (*probeData, erro
 	// loki does not support joins
 	streams := s.extractLogs(t, logs.Bytes(), logLabels)
 
-	return &probeData{ts: ts, streams: streams, tenantId: s.check.TenantId}, err
+	return &probeData{ts: ts, streams: streams, tenantId: s.check.TenantId}, nil
+}
+
+func getProbeMetrics(ctx context.Context, prober prober.ProbeFn, target string, module *bbeconfig.Module, checkInfoLabels map[string]string, summaries map[uint64]prometheus.Summary, histograms map[uint64]prometheus.Histogram, logger kitlog.Logger) (bool, []*dto.MetricFamily, error) {
+	registry := prometheus.NewRegistry()
+
+	success := runProber(ctx, prober, target, module, registry, checkInfoLabels, logger)
+
+	mfs, err := registry.Gather()
+	if err != nil {
+		return success, nil, fmt.Errorf(`extracting data from blackbox-exporter: %w`, err)
+	}
+
+	registry = prometheus.NewRegistry()
+
+	if err := getDerivedMetrics(mfs, summaries, histograms, registry); err != nil {
+		return success, nil, fmt.Errorf(`getting derived metrics: %w`, err)
+	}
+
+	dmfs, err := registry.Gather()
+	if err != nil {
+		return success, nil, fmt.Errorf(`extracting derived metrics: %w`, err)
+	}
+
+	mfs = append(mfs, dmfs...)
+
+	return success, mfs, nil
+}
+
+func runProber(ctx context.Context, prober prober.ProbeFn, target string, module *bbeconfig.Module, registry *prometheus.Registry, checkInfoLabels map[string]string, logger kitlog.Logger) bool {
+	start := time.Now()
+
+	_ = level.Info(logger).Log("msg", "Beginning check", "type", module.Prober, "timeout_seconds", module.Timeout.Seconds())
+
+	checkCtx, cancel := context.WithTimeout(ctx, module.Timeout)
+	defer cancel()
+
+	success := prober(checkCtx, target, *module, registry, logger)
+
+	duration := time.Since(start).Seconds()
+
+	probeSuccessGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_success",
+		Help: "Displays whether or not the probe was a success",
+	})
+	probeDurationGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_duration_seconds",
+		Help: "Returns how long the probe took to complete in seconds",
+	})
+	smCheckInfo := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name:        "sm_check_info",
+		Help:        "Provides information about a single check configuration",
+		ConstLabels: checkInfoLabels,
+	})
+
+	registry.MustRegister(probeSuccessGauge)
```

Comment:
as per docs 
> MustRegister panics if an error occurs

IIUC this will crash the agent process, do we want to panic in this case?

Reply:
I think it was like that before. The reasoning probably is that it won't panic because we know that that metric can be added: basically that the metric hasn't been registered before, that the labels are consistent, etc. If it ever panics, we will catch it during testing, never during production.